datumaro/components/project.py

```diff
@@ -2,486 +2,2399 @@
+        self._data.remove(name)
+
+    def remove_stage(self, target: str, name: str):
+        assert name not in {self.BASE_STAGE}, "Can't remove a default stage"
+
+        target = self._data[target]
+        idx = find(enumerate(target.stages), lambda e: e[1].name == name)
+        if idx is None:
+            raise KeyError("Can't find stage '%s'" % name)
+        target.stages.remove(idx)
+
+    def add_transform_stage(self, target: str, transform: str,
+            params: Optional[Dict] = None, name: Optional[str] = None):
+        if not transform in self._tree.env.transforms:
+            raise KeyError("Unknown transform '%s'" % transform)
+
+        return self.add_stage(target, {
+            'type': BuildStageType.transform.name,
+            'kind': transform,
+            'params': params or {},
+        }, name=name)
+
+    def add_inference_stage(self, target: str, model: str,
+            params: Optional[Dict] = None, name: Optional[str] = None):
+        if not model in self._tree._project.models:
+            raise KeyError("Unknown model '%s'" % model)
+
+        return self.add_stage(target, {
+            'type': BuildStageType.inference.name,
+            'kind': model,
+            'params': params or {},
+        }, name=name)
+
+    def add_filter_stage(self, target: str, expr: str,
+            params: Optional[Dict] = None, name: Optional[str] = None):
+        params = params or {}
+        params['expr'] = expr
+        return self.add_stage(target, {
+            'type': BuildStageType.filter.name,
+            'params': params,
+        }, name=name)
+
+    def add_convert_stage(self, target: str, format: str,
+            params: Optional[Dict] = None, name: Optional[str] = None):
+        if not self._tree.env.is_format_known(format):
+            raise KeyError("Unknown format '%s'" % format)
+
+        return self.add_stage(target, {
+            'type': BuildStageType.convert.name,
+            'kind': format,
+            'params': params or {},
+        }, name=name)
 
-        # override with our items, fallback to existing images
-        if own_source is not None:
-            log.debug("Loading own dataset...")
-            for item in own_source:
-                existing_item = subsets.setdefault(
-                        item.subset, self.Subset(self, item.subset)). \
-                    items.get(item.id)
-                if existing_item is not None:
-                    item = item.wrap(path=None,
-                        image=ExactMerge.merge_images(existing_item, item))
+    @staticmethod
+    def make_target_name(target: str, stage: Optional[str] = None) -> str:
+        if stage:
+            return '%s.%s' % (target, stage)
+        return target
 
-                subsets[item.subset].items[item.id] = item
+    @classmethod
+    def split_target_name(cls, name: str) -> Tuple[str, str]:
+        if '.' in name:
+            target, stage = name.split('.', maxsplit=1)
```

Comment:
I more or less agree about the ambiguity. The situation is the following:
1. 'x' (a source name or project).
1. 'x.y' (above + stage name)

Both can be [build] target names and 1 is what is called a build target. Currently, there is no specific term for the "build target basename", but maybe "base" could be the one. But it's still a build target name after split. I don't want to narrow the 1 only to sources, because it would be an artificial limitation with no real problems behind.

Reply:
Well, okay.

FWIW:

> Introducing such data class requires putting it to interfaces

That was indeed my intention. Ideally, we would parse the string into a `BuildTargetName` as soon as it would be received from the user and then pass this object around instead of a string.